src/nvim/api/vim.c

```diff
@@ -990,6 +990,42 @@ void nvim_set_option(uint64_t channel_id, String name, Object value, Error *err)
   set_option_to(channel_id, NULL, SREQ_GLOBAL, name, value, err);
 }
 
+/// Echo a message.
+///
+/// @param str Message
+/// @param opts Optional parameters. Keys:
+///          - hl_group: name of the highlight group used to highlight.
+///          - history: if true, add to |message-history|.
+void nvim_echo(String str, Dictionary opts, Error *err)
```

Comment:
I would like to eventually support full highlighting similar to virt text i e  `[["text", hl_1], [" more text", hl_2]]` , not sure if we should prepare for that here or better to make a separate function.

Reply:
> I would like to eventually support full highlighting

Indeed, that's sound good.

> not sure if we should prepare for that here or better to make a separate function.

Though I'm not sure, we might be better not to make many echo api.
We have already many commands to echo messages.

Reply:
I have the following options.
IMO, the first one seems to be a good balance between extensibility and ease of use.

```lua
vim.api.nvim_echo({{"msg", "WarningMsg"}}, {}) -- default {history = true}
```
or
```lua
vim.api.nvim_echo({{"msg", "WarningMsg"}}, true) -- history is mandatory arg
```
or
```lua
vim.api.nvim_echo({{"msg", "WarningMsg"}}, true, {}) -- opts (Reserved for future use)
```

Reply:
I'm leaning toward the last option (we want an option dict regardless, i e for more metadata for external UI:s later on).